### PR TITLE
iCade fixes and 8Bitdo controller support

### PR DESCRIPTION
**Pull request by pkamb**
This Pull Request brings in the commits added by @DanielFontes on his fork to address iCade and 8Bitdo controller support in issue #269.

I rebased those changes on top of the latest Provenance master and resolved the rebase conflicts.

https://github.com/jasarien/Provenance/issues/269

https://github.com/DanielFontes/Provenance

Submitting this PR so as to bring those changes into the main Provenance repo.

NOTE: I have not personally evaluated or tested these changes. Opening this PR for visibility, as I was surprised and happy to see that 8Bitdo support was added! I can and will test once my 8Bitdo controller arrives in the mail.

### CODE REVIEW

**Comment by jasarien** [comment]
Nice work, once it's been tested let me know and I'll merge it.

**Comment by Hanterdro** [comment]
I've tried it. For me it's not working. I've set my 8BITDO SFC30 in iCade mode (tested with Firmware 2.68 and 2.63). The Apple TV pairs with the controller but Provenance does not find the iCade controller when I press an key in settings like in this video manual on 1:25 https://www.youtube.com/watch?v=zD-XnZTtN0M

File: Provenance/Controller/iCade/PViCadeController.h [hunk 1]
```diff
@@ -7,11 +7,12 @@
 //
 
 #import <GameController/GameController.h>
+#import "PViCadeReader.h"
 
 @class PViCadeGamepad;
-@class PViCadeReader;
 
 @interface PViCadeController : GCController {
+        GCControllerPlayerIndex _playerIndex;
```
Comment (stuartjmoore) [inline Provenance/Controller/iCade/PViCadeController.h]: Does `playerIndex` need to be redefined? It’s already on `GCController`.

File: Provenance/Emulator/PVEmulatorViewController.m [hunk 2]
```diff
@@ -23,7 +23,9 @@
 #import "PVControllerManager.h"
 #import "PViCade8BitdoController.h"
 
-@interface PVEmulatorViewController ()
+@interface PVEmulatorViewController () {
+    UITapGestureRecognizer *_menuGestureRecognizer;
```
Comment (stuartjmoore) [inline Provenance/Emulator/PVEmulatorViewController.m]: Probably should be an `@property`.

**Comment by pkamb** [comment]
@Hanterdro Did the original fork of this work for you, or is it just this rebase that's not working? The original is on master here: https://github.com/DanielFontes/Provenance

**Comment by appletv4** [comment]
@pkamb how do i test this?

do i use this master fork ?
https://github.com/pkamb/Provenance

**Comment by pkamb** [comment]
This PR's branch is on my fork at `icade-fixes-via-DanielFontes`

https://github.com/pkamb/Provenance/tree/icade-fixes-via-DanielFontes

**Comment by felixLam** [comment]
I can confirm that it is working here with the 8bitdo SFC30. Had to build again as the storyboard wasn't correctly re-built. Also had to make sure to start provenance after connecting the pad. A force quit and restart fixed the intermittent connection issue.

**Comment by felixLam** [comment]
Thanks @pkamb for re-basing the branch onto master. Was about to do so myself, very timely!

**Comment by pkamb** [comment]
Great to hear that it seems to be working. I've added you as a collaborator on my repo if you'd like to push those storyboard changes to this branch.

**Comment by felixLam** [comment]
No changes were needed, only a clean all targets.

**Comment by appletv4** [comment]
I'm testing now , everything is working good with snes30 and nes30 

Safe to merge 


One thing that needs work is having 2 icade controller working as player 1 & 2, right now connecting 2 8bitdo controllers to provenance controls the same player

**Comment by Orphanmac** [comment]
I've been trying to read through most of the feeds to make sure I'm not asking the same question that someone else has but if you know where to direct me I would appreciate it. 

I have the 8bitdo nes30 controller and I want to use it with the provenance iOS version (using iCade) I had previously installed the 1.2.7 v of provenance and it didn't work so I scrapped it and installed the latest being the 1.3.1 v but that doesn't seem to work either. From what I understand the last time it worked consistently was the 1.2.5 and with some bugs on the 1.2.6. Can anyone confirm that for me and if so please direct me how to download the correct version please. Thank you.

ref issue #320 and #389 

P.S. I don't know a lot about coding or Xcode but am willing to learn so please be patient with me.

**Comment by Hanterdro** [comment]
Good new: I can confirm that restarting the app resolves the connection issue.

Bad news: Both 8BITDO SFC30 (2.68 and 2.63) have an annoying bug. If I press once the key down button the key down button is permanently active.

Also bad news: It's currently working only with one iCade controller. You can't register two controllers :-(

**Comment by Hanterdro** [comment]
The Bug with the permanently active key down appears when the language is set to German. If I change it to English everything works fine. I guess the bug will occur on other languages, too.

If I press  the Y key the the permanent key down bug is suppressed until I press the key down button again :)

**Comment by felixLam** [comment]
I think that the requirement to change the input language is mentioned in the SFC30 manual too. ;)

**Comment by jasarien** [comment]
Yes as far as I'm aware, iCade requires the language to be set to English, this is because the iCade works by sending key presses and if the language is different then the key presses it sends aren't the same as those registered in the code that listens for them.

**Comment by DanielFontes** [comment]
Wow nice! I only wanted to play some Mario with my 8bitdo on the AppleTV...

@Hanterdro, @felixLam: You can "fix" the key mapping for German layout in iCadeReaderView.m
"wdxazhujikol" for ON_STATES and "ecyqtrfnmpgv" for OFF_STATES
@jasarien: Maybe we can set the mapping automatically with [UITextInputMode currentInputMode].primaryLanguage in the future.

@appletv4: 2 players are currently not possible I think because every 8bitdo controller simulates the same keys and we do not know which controller sends the keys.

**Comment by Hanterdro** [comment]
@DanielFontes 
Yeah thanks. The changes in `iCadeReaderView.m` fixed the issue!
It would be nice if that mapping would be set automatically in the future 👍 

I dont know much about C# or ios/tvos or iCade programming. But I don't understand why two controllers are not possible. We have technically two "keyboards". Why does it matter that the 8bito controllers simulate the same keys? For player 1 you just have to listen to keyboard 1 and for player two to keyboard 2.

However, great work. This pull Request is a big step in a good direction :)

**Comment by DanielFontes** [comment]
@Hanterdro I only get an event with the char sent but no information about the sender itself. If there was an option to switch one controller into uppercase mode then it would be easy to distinguish between the two controllers.

**Comment by Burnsoft** [comment]
i;ve tested Peter's fork here https://github.com/pkamb/Provenance/tree/icade-fixes-via-DanielFontes 
with SFC30 on latest firmware (2.70) 
all working great! thanks so much

**Comment by justindhill** [comment]
Not sure how @appletv4 found this to be working fine on the SNES30. I found that the A/B/X/Y mappings were wrong and had to flip them around. Perhaps the NES30 has different mappings to the iCadeStates?

Anyway, will submit a PR adding another iCade Controller option to differentiate between the current mappings (presumably NES30) and the SNES30.

**Comment by appletv4** [comment]
when building the project i get a lot of yellow warnings telling me to update to recommended settings for stella and pico drive plus others should i do that or ignore ?

**Comment by jasarien** [comment]
Ignore them. They'll be updated in a later commit.

**Comment by appletv4** [comment]
@justindhill the snes30  and nes30 in provenance has the same button mapping as the mfi controllers

**Comment by justindhill** [comment]
I wasn't sure about your comment regarding the controls for Super Mario Bros NES, but this Stack Exchange question suggests you have it mixed up. In addition, I'm quite sure the controls are mapped incorrectly at least for SNES30, as the controls in Legend of Zelda are marked unambiguously.

**Comment by appletv4** [comment]
@justindhill the way I describe it above in my post is the correct way of playing mario or any other nes/snes game  on a gamepad with 4 buttons like a dualshock , nimbus , or the snes30

See here 
https://github.com/jasarien/Provenance/pull/231

**Comment by justindhill** [comment]
Ok, fair. Then the mapping for playing NES games on a SNES controller are correct. However, this has the side effect of making the mappings for playing SNES games on an SNES controller wrong, which isn't acceptable. We need to make the mappings different when playing these two types of games.

**Comment by appletv4** [comment]
to me they are mapped correctly on snes games also

in super mario world snes game for example 

on the snes30 controller the buttons do this

Y = Run

B = Jump

A = spin jump

X = dont remember

triggers
L = screen moves left
R = screen moves right


to me this plays correct, what is it that you want to change or find wrong ?

**Comment by justindhill** [comment]
Hmm, my apologies. I was wrong; the current mapping is correct. I was confused by the iCade nomenclature for buttons. In PViCade8BitdoController, pressing X triggers a code path that calls `[[weakSelf.iCadeGamepad buttonY] buttonPressed];` and pressing A triggers a code path that calls `[[weakSelf.iCadeGamepad buttonB] buttonPressed];`. This appears wrong on the surface, but actually produces the desired behavior.

**Comment by appletv4** [comment]
Is it just me or do these 8bitdo controllers have less lag than MFI controllers in provenance?
They feel more responsive , i game on a PlayStation tv /vitatv retroarch with a dualshock 3 and it feels like that

**Comment by iTKirill** [comment]
Its NOT WORK in SFC 30 and NES30 PRO!!!!

**Comment by ejuo** [comment]
@DanielFontes, @jasarien 
8Bitdo has Mode2 bluetooth keyboard (start+B for palyer 1 and start + B +R fro player 2). Then 2 controllers send different chars, and iOS support simultaneous connection of 2 keyboards. Can we include this mapping in provenance?

**Comment by maddox** [comment]
FYI, firmware 4.0 has removed iCade mode. You'll need to flash 3.0 onto your controller to get it to work with the Apple TV. 

This means no interoperability with Switch and Apple TV 😞

**Comment by jasarien** [comment]
How long has firmware 4 been in the wild?

**Comment by maddox** [comment]
May 23rd - http://www.nintendolife.com/news/2017/05/8bitdo_pads_updated_with_nintendo_switch_support_just_in_time_for_ultra_street_fighter_ii

Here's the full 4.0 changelog:

>Firmware 4.0 represents a significant update to all 8Bitdo controllers. It changes the way they operate and offers additional support. 
>
>1. There are now four unique modes:
>Android (D-Input, default mode) Hold START for 1 second, LEDs will blink once per cycle. 
>Windows (X-Input): Hold START + X for 1 second, LED will blink twice per cycle.
>macOS: Hold START + A for 1 second, LED will blink three times per cycle
>Nintendo Switch: Hold START + Y for 1 second, LED will blink four times per cycle. 
>Note 
>- When connecting a SNES30 to a Nintendo Switch in Switch Mode, DOWN+SELECT = Switch HOME button. 
>- iCade mode and Keyboard mode are no longer active in Firmware 4.0, please download and install Legacy firmware for this functionality 
>- Download the 4.0 manual from support.8bitdo.com for more details.
>

It's unfortunate because they removed keyboard mode as well, so even hacking support for keys is gone too :(

**Comment by jasarien** [comment]
So with the latest firmware, 8bitdo controllers are effectively incompatible with iOS or tvOS?

**Comment by maddox** [comment]
Yuuuup. But they still offer `3.0` on their site and you can downgrade. That's what I just did.

I'd been waiting to try this controller (SFC30) with Provenance for ages. I finally get one and this happened. lol.

**Comment by justindhill** [comment]
You should still be able to use them in BT keyboard mode, right? Or do I
misunderstand?

**Comment by maddox** [comment]
>iCade mode and Keyboard mode are no longer active in Firmware 4.0, please download and install Legacy firmware for this functionality

**Comment by maddox** [comment]
Here's the new manual with the new modes:

http://download.8bitdo.com/Manual/Controller/SNES30+SFC30/SNES30+SFC30_Manual_V4.pdf

**Comment by justindhill** [comment]
The quote didn't come through in my email. :(

**Comment by appletv4** [comment]
I only see 4.0 firmware to download  http://support.8bitdo.com


Anybody have links for 3.0

**Comment by appletv4** [comment]
It might be cause I'm on my phone  that I can't see the download link 
I'll check when I get home

**Comment by maddox** [comment]
> It might be cause I'm on my phone that I can't see the download link

This is why. I just checked on my phone and the legacy links are hidden. They're there on your desktop though.

**Comment by maddox** [comment]
I just sent an email asking if they'd reconsider adding iCade support back. I was really looking forward to having iCade and Switch modes on this controller, but now the modes are separated by firmwares, lol. I'm not sure if they're limited in space and needed to ditch iCade mode to make room for Switch mode (which would make perfect economic sense), but it'd be pretty cool if we could pick which modes we want in the firmware. Having regular HID, iCade, and Switch would be the best of all worlds.

What makes this controller so great is that just by having different modes, it means you can keep the controller paired to different devices at the same time. It's great.

**Comment by maddox** [comment]
response:

>Hello there!
>
>Thanks for your big support to 8bitdo. If we have any plan we will public on our FaceBook page. Please follow our page to know more information.
>
>Thank you so much!

¯\(°_o)/¯

**Comment by appletv4** [comment]
This looks like the perfect provenance controller 
https://www.youtube.com/watch?v=ZLWbCcRSkIk

But it won’t have the old 3.0 firmware that works with provenance ?

**Pull request merged**